File: tests/basic/loop_fission.ml
open Optitrust
open Target

let _ = Flags.check_validity := true

let _ = Run.doc_script_cpp (fun _ ->

  !! Loop_basic.fission [tBefore; cFor "i"; sInstr "y"];

)

"
#include \"../../include/optitrust.h\"

int main() {
  __pure();

  int x;
  int y;
  for (int i = 0; (i < 5); i++) {
    __sequentially_reads(\"x ~> Cell, y ~> Cell\");
    x + 0;
    y + 0;
  }
}
"

let _ = Run.script_cpp ( fun _ ->
(*
  !! Trace.failure_expected (fun () ->
    Loop_basic.fission [tBefore; sInstr "t[i] +="];
  );
*)
  !! Loop_basic.fission [tAfter; sInstr "t[i] +="];
  !! Loop_basic.fission [tBefore; cVarDef "z"];

  !! Loop_basic.fission [tAfter; sInstr "MFREE1(5, m1);"];

)

File: tests/basic/loop_fission.cpp
#include "../../include/optitrust.h"

int* const t;
int* const u;
const int n;

int main() {
  __modifies(
    "Group(range(1, n, 1), fun i -> &t[i] ~> Cell),"
    "Group(range(1, n, 1), fun i -> &u[i] ~> Cell)");

  for (int i = 1; i < n; i++) {
    __modifies(
      "&t[i] ~> Cell,"
      "&u[i] ~> Cell");
    int a = i;
    t[i] += a;
    int b = i;
    u[i] += b;
  }

  for (int i = 0; (i < 5); i++) {
    __pure();
    int y = i;
    int z = i;
  }

  for (int i = 0; i < 3; i++) {
    __pure();
    int* const m1 = (int* const) MALLOC1(5, sizeof(int));
    MFREE1(5, m1);
    int* const m2 = (int* const) MALLOC1(5, sizeof(int));
    MFREE1(5, m2);
  }

  // TODO: return 0;
}
